services/ip_manager/src/main/java/com/futurewei/alcor/ipmanager/controller/Ipv4AddrController.java

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2019 The Alcor Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+        you may not use this file except in compliance with the License.
+        You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+        Unless required by applicable law or agreed to in writing, software
+        distributed under the License is distributed on an "AS IS" BASIS,
+        WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+        See the License for the specific language governing permissions and
+        limitations under the License.
+*/
+
+package com.futurewei.alcor.ipmanager.controller;
+import com.futurewei.alcor.ipmanager.entity.Ipv4AddrState;
+import com.futurewei.alcor.ipmanager.http.Ipv4AddrRequest;
+import com.futurewei.alcor.ipmanager.http.Ipv4AddrRangeRequest;
+import com.futurewei.alcor.ipmanager.http.Ipv4AddrRequestBulk;
+import com.futurewei.alcor.ipmanager.http.status.Ipv4AddrInvalidException;
+import com.futurewei.alcor.ipmanager.http.status.Ipv4AddrRangeInvalidException;
+import com.futurewei.alcor.ipmanager.http.status.Ipv4AddrStateInvalidException;
+import com.futurewei.alcor.ipmanager.http.status.SubnetIdInvalidException;
+import com.futurewei.alcor.ipmanager.service.Ipv4AddrService;
+import com.futurewei.alcor.utils.Ipv4AddrUtil;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.Map;
+
+
+@RestController
+public class Ipv4AddrController {
+    @Autowired
+    Ipv4AddrService ipv4AddrService;
+
+    private void checkSubnetId(String subnetId) throws SubnetIdInvalidException {
```

Comment:
When you get a chance, please include the design doc in the PR as well. Thanks.

Reply:
Can you update concurrency handling section? When multiple clients try to allocate IPs from the same subnet, it could potentially get the same IPs. how do you handle this situation?